### PR TITLE
Avoid large blocks on forced compaction of blocks ingester

### LINKED ISSUE #3308: Blocks: Compaction (idle/force) should respect configured time range for blocks.
**Describe the bug**
When Cortex ingester is [idle- or force-compacting the block](https://github.com/cortexproject/cortex/blob/e26519995ae889d0f7a187a1e74dde99571fa39d/pkg/ingester/ingester_v2.go#L1219), it uses full time-range of the head as input parameter to compaction. This doesn't respect configured block range, and is especially bad when head contains data spanning several days or even weeks – it would produce huge block. We should fix that.

**Expected behavior**
Compaction (idle/force) should respect configured time range for blocks.

**Storage Engine**
- [x] Blocks
- [ ] Chunks

### PR DESCRIPTION
**Pull request by codesome**
**What this PR does**:

Creates smaller blocks out of Head during idle and forced compaction.

**Which issue(s) this PR fixes**:
Fixes #3308

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ingester/ingester_v2.go [hunk 1]
```diff
@@ -1216,8 +1217,30 @@ func (i *Ingester) compactionLoop(ctx context.Context) error {
...
+			return err
+		}
+	}
+	return userDB.CompactHead(tsdb.NewRangeHead(h, h.MinTime(), h.MaxTime()))
+}
+
 // Compacts all compactable blocks. Force flag will force compaction even if head is not compactable yet.
 func (i *Ingester) compactBlocks(ctx context.Context, force bool) {
+	i.TSDBState.compactBlocksMtx.Lock()
+	defer i.TSDBState.compactBlocksMtx.Unlock()
```
Comment (codesome) [inline pkg/ingester/ingester_v2.go:1242]: As the access to the mint and maxt of Head and then compaction is not atomic, we need this protection

File: CHANGELOG.md [hunk 2]
```diff
@@ -92,6 +92,7 @@
 * [BUGFIX] Use a valid grpc header when logging IP addresses. #3307
 * [BUGFIX] Fixed the metric `cortex_prometheus_rule_group_duration_seconds` in the Ruler, it wouldn't report any values. #3310
 * [BUGFIX] Fixed gRPC connections leaking in rulers when rulers sharding is enabled and APIs called. #3314
+* [BUGFIX] Block Storage: Avoid creating large blocks on forced compaction and when TSDB is idle. #3344
```
Comment (pracucci) [inline CHANGELOG.md:95]:
```suggestion
* [BUGFIX] Blocks storage: Avoid creating blocks larger than configured block range period on forced compaction and when TSDB is idle. #3344
```

File: pkg/ingester/ingester_v2.go [hunk 3]
```diff
@@ -1216,8 +1217,30 @@ func (i *Ingester) compactionLoop(ctx context.Context) error {
 	return nil
 }
 
+// compactHead compact the Head block at 2h intervals,
+// avoiding a single huge block.
+func compactHead(userDB *userTSDB) error {
```
Comment (pracucci) [inline pkg/ingester/ingester_v2.go:1222]: What do you think to move it to `userTSDB.compactHead()`?

File: pkg/ingester/ingester_v2.go [hunk 4]
```diff
@@ -1216,8 +1217,30 @@ func (i *Ingester) compactionLoop(ctx context.Context) error {
...
+	// The Compact() will produces multiple 2h blocks if Head has a lot of data.
```
Comment (pracucci) [inline pkg/ingester/ingester_v2.go:1223]: Do not hardcode "2h" in the comment because it depends on the configured block range period

File: pkg/ingester/ingester_v2.go (hunk 1 above)
Comment (pracucci) [reply pkg/ingester/ingester_v2.go:1242]: I would make it per TSDB, what do you think?

If yes, then we need to wrap the call to `userDB.Compact()` to take the lock there too. Moving in this direction, to make the code more clear, I would then embed `*tsdb.DB` with a name in `type userTSDB struct` so that we explicitly wrap the `tsdb.DB` functions we need to call.

**Review by pracucci — COMMENTED** [review]
Thanks @codesome ! Good work, LGTM. I left few minor comments, mostly design. I would also be glad if you could add a unit test: we already have `TestIngester_flushing` and, if I'm not missing anything, could be easy to add another test case where you push samples spanning over different days and make sure it generates 2 output blocks.

File: pkg/ingester/ingester_v2.go [hunk 5]
```diff
@@ -1216,8 +1217,30 @@ func (i *Ingester) compactionLoop(ctx context.Context) error {
...
+	if err := userDB.Compact(); err != nil {
+		return err
+	}
+	twentyFourHrs := 24 * time.Hour.Milliseconds()
+	h := userDB.Head()
+	if (h.MinTime()/twentyFourHrs)*twentyFourHrs != (h.MaxTime()/twentyFourHrs)*twentyFourHrs {
```
Comment (pracucci) [inline pkg/ingester/ingester_v2.go:1229]: Shouldn't we check the configured `TSDBConfig.BlockRanges` and compact the head if mint and maxt don't belong to the same range period?

File: pkg/ingester/ingester_v2.go (hunk 1 above)
Comment (codesome) [reply pkg/ingester/ingester_v2.go:1242]: Makes sense to make it per TSDB

File: pkg/ingester/ingester_v2.go (hunk 5 above)
Comment (codesome) [reply pkg/ingester/ingester_v2.go:1229]: That would be a better one

File: pkg/ingester/ingester_v2.go (hunk 1 above)
Comment (codesome) [reply pkg/ingester/ingester_v2.go:1242]: Taking a look now, and I think we don't need to make `*tsdb.DB` with a name right now. We don't need a wrapper around `Compact()`, but rather `CompactHead`, but the wrapper wont match the original `CompactHead` function signature. So I am thinking it to keep it simple and have `userDB.compactHead()`. WDYT? Embeding `*tsdb.DB` with a name sounds fine too if that would help later.

Comment (codesome) [reply pkg/ingester/ingester_v2.go:1242]: Oh I see now, it will be easier to use the lock if we also wrap `Compact()`. I will embed the tsdb.DB as a name in the struct.

Comment (pracucci) [reply pkg/ingester/ingester_v2.go:1242]: Embedding `tsdb.DB` with a name may help to avoid misuses, because functions must be explicitly wrapped.

**Comment by codesome** [comment]
PR is ready for another review. I have added a unit test and refactored the code as discussed.

File: pkg/ingester/ingester_v2.go [hunk 6]
```diff
@@ -48,12 +48,13 @@ type Shipper interface {
...
-	activeSeries   *ActiveSeries
-	seriesInMetric *metricCounter
-	limiter        *Limiter
+	db               *tsdb.DB
+	userID           string
+	refCache         *cortex_tsdb.RefCache
+	activeSeries     *ActiveSeries
+	seriesInMetric   *metricCounter
+	limiter          *Limiter
+	compactBlocksMtx sync.Mutex
```
Comment (pstibrany) [inline pkg/ingester/ingester_v2.go:57]: Typically mutex is on top of the block of variables that it protects. This mutex is used differently though. I would suggest adding a comment about it.

File: pkg/ingester/ingester_v2.go [hunk 7]
```diff
@@ -66,14 +67,63 @@ type userTSDB struct {
...
+	defer u.compactBlocksMtx.Unlock()
+
+	return u.db.Compact()
+}
+
+// compactHead compacts the Head block at specified block durations avoiding a single huge block.
+func (u *userTSDB) compactHead(blockDuration int64) error {
+	u.compactBlocksMtx.Lock()
+	defer u.compactBlocksMtx.Unlock()
+	// The Compact() will produces multiple smaller blocks if Head has a lot of data.
```
Comment (pstibrany) [inline pkg/ingester/ingester_v2.go:103]: typo: ... produce ...

File: pkg/ingester/ingester_v2.go [hunk 8]
```diff
@@ -66,14 +67,63 @@ type userTSDB struct {
...
+	h := u.Head()
+	if (h.MinTime()/blockDuration)*blockDuration != (h.MaxTime()/blockDuration)*blockDuration {
+		// Remaining data in Head spans across 2 block ranges, so we break it into 2 blocks.
+		// Block maxt is exclusive, so we do a -1 here for maxt.
+		maxt := (h.MaxTime()/blockDuration)*blockDuration - 1
+		if err := u.db.CompactHead(tsdb.NewRangeHead(h, h.MinTime(), maxt)); err != nil {
+			return err
+		}
+	}
+	return u.db.CompactHead(tsdb.NewRangeHead(h, h.MinTime(), h.MaxTime()))
```
Comment (pstibrany) [inline pkg/ingester/ingester_v2.go:116]: We have two `u.db.CompactHead` calls here, that only differ in max time. Let's use one, using maxt variable.

**Review by pstibrany — APPROVED** [review]
LGTM with some nits.

File: pkg/ingester/ingester_v2.go (hunk 6 above)
Comment (pstibrany) [reply pkg/ingester/ingester_v2.go:57]: Actually... I don't understand *why* we even need this, given that compaction can *only* be done from our own loop, and we never compact the same block concurrently.

File: pkg/ingester/ingester_v2.go (hunk 8 above)
Comment (codesome) [reply pkg/ingester/ingester_v2.go]: Not exactly, we are actually doing 2 compactions here

Comment (pstibrany) [reply pkg/ingester/ingester_v2.go]: You're right. I've introduced a bug into your PR 🤦

**Comment by pstibrany** [comment]
Based on offline discussion with @codesome, I've removed compaction mutex, since it's not needed.

Additionally, I've replaced `u.db.Compact()` call with a for-loop, since during head compaction (in `if` block), we could receive more data spanning 2 blocks again, which would again create bigger block in last head compact call. Using for-loop fixes that.

**Comment by codesome** [comment]
> Using for-loop fixes that.

Actually it does not, it's similar to calling a `Compact()`. Once you exit the loop here, you can still get more samples after that.

Additionally, a brute force loop might end up creating unaligned block since it is not aligning them like TSDB does (this brings one more thing that we will need to duplicate other checks that TSDB does and keep them in sync, maintenance burden?).

I propose keeping the `Compact()` call and instead rely on [this](https://github.com/cortexproject/cortex/issues/3282#issuecomment-714285039) for no samples during force flushing.

**Comment by codesome** [comment]
> Actually it does not, it's similar to calling a Compact(). Once you exit the loop here, you can still get more samples after that.

Sorry, I am wrong, I see you are re-using the maxTime and not taking it from the Head. But other checks during a `Compact()` still holds.

File: pkg/ingester/ingester_v2.go [hunk 9]
```diff
@@ -66,14 +66,61 @@ type userTSDB struct {
...
+}
+
+// compactHead compacts the Head block at specified block durations avoiding a single huge block.
+func (u *userTSDB) compactHead(blockDuration int64) error {
+	h := u.Head()
+
+	minTime, maxTime := h.MinTime(), h.MaxTime()
+
+	for (minTime/blockDuration)*blockDuration != (maxTime/blockDuration)*blockDuration {
+		// Data in Head spans across multiple block ranges, so we break it into 2 blocks.
```
Comment (codesome) [inline pkg/ingester/ingester_v2.go:102]:
```suggestion
		// Data in Head spans across multiple block ranges, so we break it here.
```

**Review by codesome — COMMENTED** [review]
LGTM! This is more robust against what Peter mentioned above. Ignore my previous comment.

File: pkg/ingester/ingester_v2_test.go [hunk 10]
```diff
@@ -1815,6 +1815,48 @@ func TestIngester_flushing(t *testing.T) {
...
 			},
 		},
+
+		"noBlockSpanningOverADayOnFlush": {
+			setupIngester: func(cfg *Config) {
+				cfg.BlocksStorageConfig.TSDB.FlushBlocksOnShutdown = false
+			},
+
+			action: func(t *testing.T, i *Ingester, m *shipperMock) {
+				// Pushing 5 samples, spanning over 3 days.
```
Comment (pracucci) [inline pkg/ingester/ingester_v2_test.go:1825]: > spanning over 3 days

Aren't they over 2 days? However I believe the test would be even stronger if samples would span 3 days and you would expect 3 blocks in output.

File: pkg/ingester/ingester_v2_test.go [hunk 11]
```diff
@@ -1815,6 +1815,48 @@ func TestIngester_flushing(t *testing.T) {
 				m.AssertNumberOfCalls(t, "Sync", 1)
 			},
 		},
+
+		"noBlockSpanningOverADayOnFlush": {
```
Comment (pracucci) [inline pkg/ingester/ingester_v2_test.go:1819]: > noBlockSpanningOverADayOnFlush

What does "noBlock" mean here? Also could you write it in plain english, plese?

File: pkg/ingester/ingester_v2_test.go [hunk 12]
```diff
@@ -1815,6 +1815,48 @@ func TestIngester_flushing(t *testing.T) {
...
+				pushSingleSampleAtTime(t, i, 25*time.Hour.Milliseconds())
+				pushSingleSampleAtTime(t, i, 26*time.Hour.Milliseconds())
+
+				// Nothing shipped yet.
+				m.AssertNumberOfCalls(t, "Sync", 0)
+
+				i.FlushHandler(httptest.NewRecorder(), httptest.NewRequest("POST", "/flush", nil))
+
+				// Flush handler only triggers compactions, but doesn't wait for them to finish. Let's wait for a moment, and then verify.
+				time.Sleep(3 * time.Second)
```
Comment (pracucci) [inline pkg/ingester/ingester_v2_test.go:1838]: Instead of `time.Sleep()` you could use `util.Poll()` checking the number of times `Sync()` has been called. This would make test faster and avoid flaky test if it takes more than 3s for any reason.

**Review by pracucci — APPROVED** [review]
Good job! LGTM, thanks! I just left a couple of minor nits.

File: pkg/ingester/ingester_v2_test.go (hunk 10 above)
Comment (codesome) [reply pkg/ingester/ingester_v2_test.go:1825]: > However I believe the test would be even stronger if samples would span 3 days and you would expect 3 blocks in output.

Comment (codesome) [reply pkg/ingester/ingester_v2_test.go:1825]: > However I believe the test would be even stronger if samples would span 3 days and you would expect 3 blocks in output.

It will depend on the actual samples, because by default it will try to create 2h blocks and not over a day. So a test of creating smaller blocks + breaking the block when it spans 2 day might be fine. I will update the test and maybe have 3 days itself anyway.

File: pkg/ingester/ingester_v2_test.go (hunk 11 above)
Comment (codesome) [reply pkg/ingester/ingester_v2_test.go:1819]: I will update it :P
